File: sdk/python/virtengine/modules/veid.py
"""VEID module helpers."""

from __future__ import annotations

from typing import List, Optional

import grpc

from virtengine.veid.v1 import query_pb2, query_pb2_grpc


class VEIDModule:
    """VEID module queries."""

    def __init__(self, channel: grpc.aio.Channel):
        self._query = query_pb2_grpc.QueryStub(channel)

    async def identity(self, account_address: str) -> query_pb2.QueryIdentityResponse:
        request = query_pb2.QueryIdentityRequest(account_address=account_address)
        return await self._query.Identity(request)

    async def identity_record(self, account_address: str) -> query_pb2.QueryIdentityRecordResponse:
        request = query_pb2.QueryIdentityRecordRequest(account_address=account_address)
        return await self._query.IdentityRecord(request)

    async def scope(self, account_address: str, scope_id: str) -> query_pb2.QueryScopeResponse:
        request = query_pb2.QueryScopeRequest(account_address=account_address, scope_id=scope_id)
        return await self._query.Scope(request)

    async def scopes(self, account_address: str) -> query_pb2.QueryScopesResponse:
        request = query_pb2.QueryScopesRequest(account_address=account_address)
        return await self._query.Scopes(request)

    async def scopes_by_type(
        self, account_address: str, scope_type: int
    ) -> query_pb2.QueryScopesByTypeResponse:
        request = query_pb2.QueryScopesByTypeRequest(
            account_address=account_address, scope_type=scope_type
        )
        return await self._query.ScopesByType(request)

    async def approved_clients(self) -> query_pb2.QueryApprovedClientsResponse:
        request = query_pb2.QueryApprovedClientsRequest()
        return await self._query.ApprovedClients(request)

    async def params(self) -> query_pb2.QueryParamsResponse:
        request = query_pb2.QueryParamsRequest()
        return await self._query.Params(request)

File: sdk/rust/src/modules/veid.rs
use tonic::transport::Channel;

use crate::proto::virtengine::veid::v1::{
    query_client::QueryClient, QueryApprovedClientsRequest, QueryIdentityRecordRequest,
    QueryIdentityRequest, QueryParamsRequest, QueryScopeRequest, QueryScopesByTypeRequest,
    QueryScopesRequest, QueryApprovedClientsResponse, QueryIdentityRecordResponse,
    QueryIdentityResponse, QueryParamsResponse, QueryScopeResponse, QueryScopesByTypeResponse,
    QueryScopesResponse,
};
use crate::{Error, Result};

pub struct VEIDModule {
    client: QueryClient<Channel>,
}

impl VEIDModule {
    pub fn new(channel: Channel) -> Self {
        Self {
            client: QueryClient::new(channel),
        }
    }

    pub async fn identity(&mut self, account_address: &str) -> Result<QueryIdentityResponse> {
        let request = QueryIdentityRequest {
            account_address: account_address.to_string(),
        };
        self.client
            .identity(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn identity_record(
        &mut self,
        account_address: &str,
    ) -> Result<QueryIdentityRecordResponse> {
        let request = QueryIdentityRecordRequest {
            account_address: account_address.to_string(),
        };
        self.client
            .identity_record(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn scope(
        &mut self,
        account_address: &str,
        scope_id: &str,
    ) -> Result<QueryScopeResponse> {
        let request = QueryScopeRequest {
            account_address: account_address.to_string(),
            scope_id: scope_id.to_string(),
        };
        self.client
            .scope(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn scopes(&mut self, account_address: &str) -> Result<QueryScopesResponse> {
        let request = QueryScopesRequest {
            account_address: account_address.to_string(),
        };
        self.client
            .scopes(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn scopes_by_type(
        &mut self,
        account_address: &str,
        scope_type: i32,
    ) -> Result<QueryScopesByTypeResponse> {
        let request = QueryScopesByTypeRequest {
            account_address: account_address.to_string(),
            scope_type,
        };
        self.client
            .scopes_by_type(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn approved_clients(&mut self) -> Result<QueryApprovedClientsResponse> {
        let request = QueryApprovedClientsRequest {};
        self.client
            .approved_clients(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }

    pub async fn params(&mut self) -> Result<QueryParamsResponse> {
        let request = QueryParamsRequest {};
        self.client
            .params(request)
            .await
            .map(|res| res.into_inner())
            .map_err(|e| Error::Query(e.to_string()))
    }
}
